heliopy/data/mms.py

```diff
@@ -7,131 +7,1781 @@
+        offline : bool
+            If True, file information will be gathered from the local file
+            system only (i.e. no requests will be posted to the SDC).
+        optdesc : str, list
+            Optional descriptor of the data products.
+        site : str
+            Indicate that requests should be posted to the public or private
+            side of the SDC. The private side needs team log-in credentials.
+                "public" - access the public side.
+                "private" - access the private side.
+        start_date : str, `datetime.datetime`
+            Start time of data interval of interest. If a string, it must be
+            in ISO-8601 format: YYYY-MM-DDThh:mm:ss, and is subsequently
+            converted to a datetime object.
+        version : str, list
+            File version number. The format is X.Y.Z, where X, Y, and Z, are
+            the major, minor, and incremental version numbers.
+
+        Returns
+        -------
+        obj : `heliopy.data.mms.MMSDownloader`
+            An object instance to interface with the MMS Science Data Center.
+        """
+
+        # Set attributes
+        #   - Put site before level because level will auto-set site
+        #   - Put files last because it will reset most fields
+        self.site = site
+
+        self.anc_product = anc_product
+        self.data_type = data_type
+        self.dropbox_root = dropbox_root
+        self.end_date = end_date
+        self.instr = instr
+        self.level = level
+        self.mirror_root = mirror_root
+        self.mode = mode
+        self.offline = offline
+        self.optdesc = optdesc
+        self.sc = sc
+        self.start_date = start_date
+        self.version = version
+
+        self.files = files
+
+        # Setup download directory
+        if data_root is None:
+            data_root = pathlib.Path(heliopy.config['download_dir']) / 'mms'
+
+        self.data_root = data_root
+        self._sdc_home = 'https://lasp.colorado.edu/mms/sdc'
+        self._info_type = 'download'
+
+        # Create a persistent session
+        self._session = requests.Session()
+        self._session.auth = (sdc_username, sdc_password)
+
+    def __str__(self):
+        return self.url()
+
+    # https://stackoverflow.com/questions/17576009/python-class-property-use-setter-but-evade-getter
+    def __setattr__(self, name, value):
+        """Control attribute values as they are set."""
+
+        # TYPE OF INFO
+        #   - Unset other complementary options
+        #   - Ensure that at least one of (download | file_names |
+        #     version_info | file_info) are true
+        if name == 'anc_product':
+            self.data_type = 'ancillary'
+
+        elif name == 'data_type':
+            if 'gls_selections' in value:
+                if value[15:] not in ('mp-dl-unh',):
+                    raise ValueError('Unknown GLS Selections type.')
```

Comment:
Would be helpful to list the valid values in the error message (or in this case what a valid value looks like)